Wrappers/Python/cil/optimisation/functions/KullbackLeibler.py

```diff
@@ -1,31 +1,239 @@
+# limitations under the License.
 
 import numpy
 from cil.optimisation.functions import Function
 from numbers import Number
-import functools
 import scipy.special
+import logging
+
 try:
     import numba
     from numba import jit, prange
     has_numba = True
-    '''Some parallelisation of KL calls'''
+except ImportError as ie:
+    has_numba = False
+
+class KullbackLeibler(Function):
+
+    r""" Kullback Leibler
+            
+    .. math:: F(u, v)
+            = \begin{cases} 
+            u \log(\frac{u}{v}) - u + v & \mbox{ if } u > 0, v > 0\\
+            v & \mbox{ if } u = 0, v \ge 0 \\
+            \infty, & \mbox{otherwise}
+            \end{cases}  
+
+    where the :math:`0\log0 := 0` convention is used.  
+              
+    **kwargs
+        b : DataContainer, non-negative      
+            Translates the function at point :code:`b`.
+        mask : DataContainer, default = None
+               Mask for the data :code:`b`
+        eta : DataContainer, default = array of zeros
+              Background noise
+
+    Raises
+    ------
+    ValueError
+        If :code:`b` is None.
+    ValueError
+        If :code:`b` is not None and has negative values.        
+                                            
+    
+    Note
+    ----
+    The Kullback-Leibler function is used in practice as a fidelity term in minimisation problems where the
+    acquired data follow Poisson distribution. If we denote the acquired data with :code:`b`
+    then, we write
+    
+    .. math:: \underset{i}{\sum} F(b_{i}, (v + \eta)_{i})
+    
+    where, :math:`\eta` is an additional noise. 
+    
+    In the case of Positron Emission Tomography reconstruction :math:`\eta` represents 
+    scatter and random events contribution during the PET acquisition. Hence, in that case the KullbackLeibler
+    fidelity measures the distance between :math:`\mathcal{A}v + \eta` and acquisition data :math:`b`, where
+    :math:`\mathcal{A}` is the projection operator. This is related to `PoissonLogLikelihoodWithLinearModelForMean <http://stir.sourceforge.net/documentation/doxy/html/classstir_1_1PoissonLogLikelihoodWithLinearModelForMean.html>`_ ,
+    definition that is used in PET reconstruction in the `SIRF <https://github.com/SyneRBI/SIRF>`_ software.
+    
+       
+    Note
+    ----
+    The default implementation uses the build-in function `kl_div <https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.kl_div.html>`_ from scipy.
+    The methods of the :class:`.KullbackLeibler` are accelerated provided that `numba <https://numba.pydata.org/>`_ library is installed.
+           
+    Examples
+    --------
+    >>> from cil.optimisation.functions import KullbackLeibler 
+    >>> from cil.framework import ImageGeometry   
+    >>> ig = ImageGeometry(3,4)
+    >>> data = ig.allocate('random')
+    >>> F = KullbackLeibler(b = data)
+                        
+    """     
+
+    def __new__(cls, b, eta = None, mask = None, backend = None):
```

Comment:
should be `backend = 'numba' ` if that is the default.

Reply:
default backend is `numba`. Numba added as a CIL requirement.